### PR TITLE
[Core] Base AnalysisStage

### LINKED ISSUE #1711: Naming of Main Analysis class functions
This issue is a continuation of discussion about the names of the functions which are to be public in Main Analysis class. Initial discussion was in #1703 . 

We , me and @philbucher  added a [wiki page](https://github.com/KratosMultiphysics/Kratos/wiki/Common-Python-Interface-of-Applications-for-Users) describing these set of functions. 

As suggested in the wiki page, the analysis class needs : 
- Initialize  Initializes the analysis (called once at the beginning of the simulation)
- InitializeTimeStep Initializes the Timestep (called once at the beginning of a timestep)
- SolveTimeStep Solves the time step (can be called several times within a timestep (e.g. strong coupling in FSI) )
- FinalizeTimeStep Finalizes the Timestep (called once at the end of a timestep)
- Finalize Finalizes the analysis (called once at the end of the simulation)
- Run executes the entire analysis
- RunMainTemporalLoop Executes the entire time loop

The discussion is are these functions and their names descriptive enough for the Main Analysis class.

### PR DESCRIPTION
**Pull request by philbucher**
This PR adds the BaseClass of the `AnalysisStage`, this has been discussed in several issues and telephone conferences

Thoughts are welcome

Once this is merged it is planned that `StructuralMechanics` and `FluidDynamics` make use of it

closes #1711 and #1278 

I will update the Wiki once this is merged

### CODE REVIEW

**Review by loumalouomega — COMMENTED** [review]
The others analysis scripts should derive from this one?

**Comment by philbucher** [comment]
@loumalouomega yes

**Comment by loumalouomega** [comment]
OK, I assume that like mine (contact) derives from the one in the
structural I will not need further modifications

**Comment by philbucher** [comment]
Most probably not, in case it does I can do it I guess when I use it for the Structural

File: kratos/python_scripts/base_analysis_stage.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+        self.Initialize()
+        self.RunSolutionLoop()
+        self.Finalize()
+
+    def RunSolutionLoop(self):
+        """This function executes the solution loop of the AnalysisStage
+        It can be overridden by derived classes
+        """
+        while self.time < self.end_time:
+            self.AdvanceInTime(self.time)
```
Comment (pooyan-dadvand) [inline kratos/python_scripts/base_analysis_stage.py]: I think this part should be delegated to a solver class in another layer designed for developer as it is right now.

File: kratos/python_scripts/base_analysis_stage.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+        This function has to be implemented in deriving classes!
+        """
+        raise NotImplementedError("This function has to be implemented by derived classes")
+
+    def AdvanceInTime(self, new_time):
+        """This function prepares the database for the new time-step
+        It is designed to be called once at the beginning of a time-step
+        It typically calles the CloneTimeStep function of a ModelPart
+        """
+        pass
```
Comment (pooyan-dadvand) [inline kratos/python_scripts/base_analysis_stage.py]: It won't be better to call solver->EstimateDT() and clone time step in the base class? 

what would be the case for load increments and other relative (non transient) situations?

**Review by pooyan-dadvand — DISMISSED** [review]
Thanks for your effort in making this PR but I don't see it mature enough to be approved yet and there are some very important decisions are still to be made and are not covered here:
- Separating or not the two layers (here is not separated and I don't like it personally)
- Having different interfaces for users and developers
- I have not used two generic names like Analysis and Stage for one object. (we may use one)
- The base class should implement the "standard" operations like basic check of parameters and calling standard factories. 
- RestartSolutionStep is missing

I'm sure that most of these comments are discussed in skype meeting but as long as those discussion are not reflected here I don't have enough arguments to accept this PR. 

Some comments in lines.

**Comment by pooyan-dadvand** [comment]
I have just put a request change to prevent the merge and we will follow this also in @KratosMultiphysics/technical-committee

File: kratos/python_scripts/base_analysis_stage.py (hunk 1 above)
Comment (jcotela) [reply kratos/python_scripts/base_analysis_stage.py]: @pooyan-dadvand The goal for this method is to encapsulate the time advancement of the Kratos database with a view to coupled analysis, involving multiple analysis/stage objects. We need a way to ensure that the database is ready before we start doing things with the new time step values (for all eventual coupled sub-problems, which may or may not share the nodes: think fluid + temperature versus FSI), and this is why we decided to give a separate method for this operation, at the analysis/stage level.

**Comment by jcotela** [comment]
Regarding the one layer vs. two layer discussion: we decided to postpone it. The goal for this PR is to decide a common "outer" interface for the analysis/stage class, and leave out to each implementation to decide if an "inner" solver layer is required. Given one or more analysis/stage instances, we should be able to perform the complete simulation without having to ask the instances for their solvers.

Our agreement was that, once the analyses/stages are implemented for the main applications, we would re-evaluate if an internal solver object with a clean interface/clear responsabilities can be "separated" from the analysis/stage. If so, we go for a two-level design, otherwise we agree on a single layer.

**Comment by RiccardoRossi** [comment]
@jcotela, this morning we have been discussing about this with @pooyan-dadvand. 
We shall speak about this in the technical committee. actually i would propose to have a meeting of the @KratosMultiphysics/technical-committee this week. Guys would that be possible for you?

**Comment by maceligueta** [comment]
Ok for me on thursday or friday, not tomorrow.

**Comment by jcotela** [comment]
fine for me, let me know when is the meeting.

**Comment by pooyan-dadvand** [comment]
What about Friday 10:00?

**Comment by RiccardoRossi** [comment]
@pooyan-dadvand @KratosMultiphysics/technical-committee  for me fine friday at 10.

for the meeting I would like to discuss on:
1 - Stage
2 - MPCs in modelpart

**Comment by jcotela** [comment]
Friday at 10 is fine. @RiccardoRossi let's not make it last until lunch time...

**Comment by maceligueta** [comment]
Fine Friday at 10.

**Comment by RiccardoRossi** [comment]
Hello everyone,
     the @KratosMultiphysics/technical-committee reached a consensus on how the "AnalysisStage" should look like.

The decision is to go for a two-level design in which a solver is used as a part of the stage.
You will see that the design follows what discussed, introducing however a separation between "solver" and "stage". this implies that coupling should be implemented as new coupled solver, unless they use different phsyics completely as black box.

@KratosMultiphysics/technical-committee please take a look if what i am committing corresponds to what discussed. I completed the file by calling the processes in the correct place within the script

The rationale is that "solver" developement will require deep technical knowledge of the physics, while touching the stage should be doable by "normal users" who are not deeply into the details of the solution.

**Comment by RiccardoRossi** [comment]
Forgot to tell that we are still mantaining that the solver handles the Importing, however this is not a definitive solution and we will look again into it later on. Functionality will be unchanged, so this decision is to move forward.

**Comment by philbucher** [comment]
I removed the whitespaces that codacy complained abt
@RiccardoRossi I suggest you set your editor to delete trailing whitespaces when saving, this is quite convenient

**Review by RiccardoRossi — DISMISSED** [review]

**Comment by jcotela** [comment]
@pooyan-dadvand could you re-review this PR?

**Comment by pooyan-dadvand** [comment]
I have removed the Predict method and dismissed my review so another one can approve it.

**Comment by pooyan-dadvand** [comment]
Just to add that now I am confirm with it. 👍

**Review by RiccardoRossi — APPROVED** [review]

**Comment by philbucher** [comment]
It seems like an agreement has been reached :+1: 

Since I started this PR I would merge it this evening if none complains

**Comment by philbucher** [comment]
@maceligueta If you want to use this in dEM I suggest waiting a week until @jcotela and I use it in Fluid and Structural just in case there will be some minor modifications

**Comment by maceligueta** [comment]
Ok, I'll wait some time.

**Pull request merged**